proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -92,6 +95,7 @@ type roundTripper struct {
 	retryableClassifier fails.Classifier
 	errorHandler        errorHandler
 	routeServicesClient http.RoundTripper
+	endpointTimeout     time.Duration
 }
 
 func (rt *roundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
```

Comment:
@swetharepakula @zachgersh This implementation of `http.RoundTripper` looks like it may violate some of the assumptions stated in godoc https://golang.org/pkg/net/http/#RoundTripper, in particular but possibly not limited to:

> RoundTrip should not modify the request, except for consuming and closing the Request's Body

I'll take a look at the overall intent. If there's anything to change I'll make a proposal.

Reply:
Just below what I can comment on in this PR:

```go
if request.Body != nil {
    closer := request.Body
    request.Body = ioutil.NopCloser(request.Body)
    defer func() {
        closer.Close()
    }()
}
```

- `(*http.Request).Body` is an `io.ReadCloser`: https://golang.org/src/net/http/request.go#L181
- By wrapping it in an `ioutil.NopCloser` the Body is never closed: https://golang.org/src/io/ioutil/ioutil.go#L114
- Which is a requirement of `RoundTripper`: https://golang.org/src/net/http/client.go#L133

Reply:
@judwhite glad you found this. 

but doesn't `closer.Close()` cause the original request Body to be closed?  I think that's the idea behind the variable assignment before modifying the `request.Body`

Reply:
@rosenhouse You're right, I read it wrong. I guess I'm confused why the code isn't more like:

```go
if request.Body != nil {
    defer request.Body.Close()
}
```

I suppose this prevents the passed in `io.ReadCloser` `Body` from being closed, but without looking at the bigger picture I'm not sure why.

@rosenhouse Can you also look at https://github.com/cloudfoundry/gorouter/pull/211/files#r173833484 ?